044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+broker will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go offline
+if all volumes are unavailable.
+
+Throttling down message production on all broker nodes will protect the cluster from
+running out of disk due to replication.
+
+See also:
+* [kafka-quotas-plugin issue #28 - KIP-827 support](https://github.com/strimzi/kafka-quotas-plugin/issues/28)
+* [kafka-quotas-plugin issue #2 - Improve support for JBOD storage](https://github.com/strimzi/kafka-quotas-plugin/issues/2)
+
+## Motivation
+
+Users need better protection against running out of disk on any of their nodes as it can degrade the service
+in unpredictable ways (corrupted segment logs, affect other users of that disk etc) which could impede
+the interventions required to recover.
+
+## Proposal
+
+1. Extend the kafka-quotas-plugin so that we can choose to observe the disk usage 
+of all brokers in the cluster.
+2. Deprecate the existing local disk observations.
+3. Add new limit types, so that it is explicit what is being limited and better support heterogeneous disks
+   1. throttle if consumed bytes exceeds threshold on any volume
+   2. throttle if usable bytes less than threshold on any volume
+   3. throttle if usable ratio less than threshold on any volume
+4. Introduce an extension point in the quotas-plugin to support pluggable sources of quotas
+5. Introduce a fallback throttle factor in case we cannot retrieve the cluster state
+
+So every broker will make its own independent throttling decision based on knowledge of the volumes on all active broker nodes. If it detects that any volume in
+the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich enough to do anything 
+smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
+
+### High level changes
+To better support external sources for managing quotas this proposal introduces some new concepts to the plugin:
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an extension point where we could plug in
+future sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `capacityBytes`
+4. `usableBytes`
```

Comment:
I wonder about the naming here. Is that usable space? For me that would normally be the full capacity of the disk so it makes me curious how does it differ from capactiyByzes. Or - I guess - is it free disk space?

Reply:
This is meant to mean free/available (i.e total minus used). We followed the [naming](https://kafka.apache.org/33/javadoc/org/apache/kafka/clients/admin/LogDirDescription.html#usableBytes--) from the kafka admin api here, but maybe it's too peculiar.

How does `totalBytes`/`availableBytes`/`usedBytes` sound instead? Or `totalBytes`/`freeBytes`/`usedBytes`.

Reply:
I would probably prefer the second. But I can live with both => both seem to be quite clear and understandable. Thanks.